Minuit2 remove unnecessary assert

HDembinski opened the pull request:
This patch removes an unnecessary assert, which is probably a left-over from early debugging.

Asserts are a way to document certain runtime assumptions that the code makes. They should be only used if the code cannot run correctly if the condition is not fulfilled. This assert here does not serve that purpose. Starting the minimization iteration with an invalid initial state is fine as far as the algorithm is concerned, because it does not make use of that property.

Edit: This is not a cosmetic change, because this assert triggers an abort in one of the iminuit tests if the code is compiled in debug mode, while the test runs fine in release mode.

eguiraud commented:
@phsft-bot build please

lmoneta reviewed (approved):
I agree this assert is not needed and should be removed. 
However the initial state in general should be valid, maybe there was an error in computing the initial state  return MnSeed. 
Maybe is worth adding a warning message

HDembinski commented:
@lmoneta Good point, I am happy to add a warning message once the logging patch is merged.

I looked into the failing tests, these seem to be unrelated (something about Python).

eguiraud commented:
@HDembinski yes those are known unrelated failure, it's #6753

HDembinski commented:
@lmoneta Now that the logging patch is in, I added a warning message to replace the assert. Good to merge?

HDembinski commented:
Perhaps even an `Error` would be appropriate instead of a `Warn`. The code still works, though.

HDembinski commented:
For reference, the following iminuit test trigger this warning or the assert:
```py
from iminuit import Minuit

m = Minuit(lambda a: -a, 0)
m.migrad()
```

lmoneta commented:
Hi, 

Thank you Hans for your example, it was very useful for this case. 

I have investigate this and I think the assert should be removed and there is no need of a warning or an error.

The first time one iterates the first states comes from MnSeed and I think in that case the state is always valid (but not 100% sure)
Eventually we could add the assert or (better) an Error but not at line  232, but at line 75, when we enter in the top function VariableMetricBuilder::Minimum

What we should maybe change is  the logic around line 150. A covariance matrix is computed via MnHesse. This is done to check the edm. If the edm is still too large the minimization is continued. 
However, if Hesse fails, (for example in the test above f(x) = -x, because the second-derivatives is zero) there is no real point in continuing the minimization. The correct edm will not make sense and one should stop iterating.  
Maybe we could add a Warning there before line 146, like
```
if (!st.IsValid()) { 
        print.Warn("Invalid Hessian - exit the minimization");
        break; 
}
```

Let me know what you think 
Cheers
Lorenzo

HDembinski commented:
@lmoneta Thank you for the comprehensive reply! What you suggests sounds better indeed. I will try your suggestion in iminuit and then do the changes as you requested.

HDembinski commented:
I had a closer look in the code and can give a more detailed answer.

 > The first time one iterates the first states comes from MnSeed and I think in that case the state is always valid (but not 100% sure)

I think you are right. BasicMinimumSeed has only one ctor that allows one to set values and that always sets the fValid to true. There is no interface to change it afterwards. We could also remove the flag, it has no purpose.

> Eventually we could add the assert or (better) an Error but not at line 232, but at line 75, when we enter in the top function VariableMetricBuilder::Minimum

Ok, I implemented that.
 
> What we should maybe change is the logic around line 150. A covariance matrix is computed via MnHesse. This is done to check the edm. If the edm is still too large the minimization is continued.
> However, if Hesse fails, (for example in the test above f(x) = -x, because the second-derivatives is zero) there is no real point in continuing the minimization. The correct edm will not make sense and one should stop iterating.
> Maybe we could add a Warning there before line 146, like
> 
> ```
> if (!st.IsValid()) { 
>         print.Warn("Invalid Hessian - exit the minimization");
>         break; 
> }
> ```

Ok, I also implemented that.

HDembinski commented:
@lmoneta Is this ready to be merged now? It would be really helpful to fix the crash I am getting in iminuit...

lmoneta commented:
Yes, It is ready. Let me run the CI and afterwards I will merge it
Thank you very much for updating it!
Cheers

lmoneta commented:
@phsft-bot build  with flags -Dminuit2=On

lmoneta commented on math/minuit2/src/VariableMetricBuilder.cxx:
```diff
@@ -69,15 +69,21 @@ FunctionMinimum VariableMetricBuilder::Minimum(const MnFcn &fcn, const GradientC
    FunctionMinimum min(seed, fcn.Up());
 
+   if (seed.Parameters().Vec().size() == 0) {
+      print.Error("No free parameters.");
```
Now this adds an Error message in stressRooStats, where in some case the conditional fit are made and likelihood has no free parameters. 
I think we could just add a Warning message here.

HDembinski replied:
Ok, I changed.

HDembinski commented:
@lmoneta I changed the error to a warning, I hope we can merge now. :)

lmoneta commented:
@phsft-bot build with flags -Dminuit2=On
